Updates to display, editor and authz tag helpers

MSingh-13 opened the pull request:
Hi @DamianEdwards 

I would appreciate it if you would review and consider merging these updates into the next version of the library.

1. Updated editor and display tag helpers to accept class and style attributes.
This allows adding classes and styles to the tag helpers as normal html attributes (with IntelliSense support). The values are passed to the editor/display template as an anonymous object on the htmlAttributes property of the ViewData which is how they were passed when using editor templates in MVC5. This allowed me to reuse my existing editor/display templates from MVC5 with minimal changes. I used a modified version of the MergeHtmlAttributes extension created by Chris Pratt.
https://web.archive.org/web/20211205165309/https://cpratt.co/html-editorfor-and-htmlattributes/  (the original site is currently not accessible)

3. Updated AuthzTagHelper to handle authorizations based on User roles and permissions. 
I have updated the TagHelperPack.Sample project to demo these use cases.
a) User role is checked using the Principal.IsInRole() method provided by the framework.
b) Permission is checked in an authorization policy based on the permissions associated with the current user role.

DamianEdwards commented on samples/TagHelperPack.Sample/Startup.cs:
```diff
@@ -53,6 +53,25 @@ public void ConfigureServices(IServiceCollection services)
             });
+            options.AddPolicy("PermissionPolicy", policy =>
+            {
+                List<string> standardPermissions = new List<string>() { "ViewUsers"};
+                List<string> adminPermissions = new List<string>(standardPermissions) { "ManageUsers" };
```
```suggestion
                var standardPermissions = new List<string> { "ViewUsers" };
                var adminPermissions = new List<string>(standardPermissions) { "ManageUsers" };
```

DamianEdwards commented on src/TagHelperPack/AuthzTagHelper.cs:
```diff
@@ -49,6 +53,18 @@ public AuthzTagHelper(IAuthorizationService authz)
+    public string RequiredRole { get; set; }
+
+    /// <summary>
+    /// A permission that must be satisfied in order for the current element to be rendered. asp-authz-policy should be set as well.
```
```suggestion
    /// A permission that must be satisfied in order for the current element to be rendered. <c>asp-authz-policy</c> should be set as well.
```

DamianEdwards commented on src/TagHelperPack/DisplayTagHelper.cs:
```diff
@@ -62,6 +62,18 @@ public DisplayTagHelper(IHtmlHelper htmlHelper)
     public ViewContext ViewContext { get; set; }
 
+    /// <summary>
+    /// css class
```
```suggestion
    /// CSS class name.
```

DamianEdwards commented on src/TagHelperPack/DisplayTagHelper.cs:
```diff
@@ -62,6 +62,18 @@ public DisplayTagHelper(IHtmlHelper htmlHelper)
+    public string Class { get; set; }
+
+    /// <summary>
+    /// css style
```
```suggestion
    /// CSS inline style.
```

DamianEdwards commented on src/TagHelperPack/EditorTagHelper.cs:
```diff
@@ -55,13 +55,25 @@ public EditorTagHelper(IHtmlHelper htmlHelper)
     public ViewContext ViewContext { get; set; }
 
+    /// <summary>
+    /// css class
```
```suggestion
    /// CSS class name.
```

DamianEdwards commented on src/TagHelperPack/EditorTagHelper.cs:
```diff
@@ -55,13 +55,25 @@ public EditorTagHelper(IHtmlHelper htmlHelper)
+    public string Class { get; set; }
+
+    /// <summary>
+    /// css style
```
```suggestion
    /// CSS inline style.
```

DamianEdwards commented on src/TagHelperPack/HtmlHelperExtensions.cs:
```diff
@@ -3,6 +3,9 @@
 using Microsoft.AspNetCore.Mvc.Rendering;
 using Microsoft.AspNetCore.Html;
 using TagHelperPack;
+using System.Collections.Generic;
```
nit: order the using statements

MSingh-13 commented:
Hi @DamianEdwards.

I have incorporated your feedback and made a new commit.

MSingh-13 committed:
Updated editor and display taghelpers to accept class and style attributes.

Updated global.json
"rollForward": "latestMajor" because did not have sdk verison 7 installed.

MSingh-13 committed:
Updated AuthzTagHelper to handle authorizations based on permissions (using string resource with the specified policy) and roles (Principal.IsInRole()).

MSingh-13 committed:
Incorporated PR feedback

1. Using var to declare permissions list in Startup.cs

2. Updated the documentation text.

3. Sorted the using statements in HtmlHelperExtensions.

Co-authored-by: Damian Edwards <damian@damianedwards.com>

MSingh-13 commented:
I just realised that my commits contained my work email. Since the changes had not been merged yet I decided to change the email by doing a rebase and amend on all my previous commits. I then force pushed them.

I did not make any changes to the source code of the first 2 commits that you had reviewed a couple of days ago. The last commit incorporates the feedback from your review.

I apologise for any inconvenience that this may have caused.

MSingh-13 commented:
@DamianEdwards 

Just checking in to see if you got a chance to review the updates.

MSingh-13 committed:
Updated samples to display source for asp-authz-permission and asp-authz-role.

DamianEdwards commented on src/TagHelperPack/AuthzTagHelper.cs:
```diff
@@ -12,13 +12,17 @@ namespace TagHelperPack;
     private const string AspAuthzAttributeName = "asp-authz";
     private const string AspAuthzPolicyAttributeName = "asp-authz-policy";
+    private const string AspAuthzRoleAttributeName = "asp-authz-role";
+    private const string AspAuthzPolicyPermissionAttributeName = "asp-authz-permission";
```
I'm not super comfortable with this name as it's too specific to the authorization handler implementation in the sample. The "permission" isn't a strongly-typed primitive/concept of the authorization system but instead is passed through the `object resource` parameter which is intended as an untyped escape-hatch to pass arbitrary information from the authorization site to the authorization handler. Perhaps your change could be updated to reflect that, e.g. change the name to `asp-authz-resource` and type it as `object`.

MSingh-13 replied:
I have renamed permission to resource.

See commit 1ae3e047d575d413c059996700f7b7be13ab3690

DamianEdwards commented on src/TagHelperPack/AuthzTagHelper.cs:
```diff
@@ -85,23 +101,45 @@ public override async Task ProcessAsync(TagHelperContext context, TagHelperOutpu
-
             showOutput = authorized;
         }
+        else if (!string.IsNullOrEmpty(RequiredRole))
```
Add a guard to ensure invalid combinations of attributes aren't being set and throw an `InvalidOperationException` in that case, e.g. if both a required role and policy are set.

MSingh-13 replied:
As suggested, I have added a condition to guard against RequiredRoles and RequiredPolicy being set at the same time.
See commit  fbdab700828f8b71e0d06cc97efe55193f772411

DamianEdwards commented on src/TagHelperPack/HtmlHelperExtensions.cs:
```diff
@@ -95,4 +98,58 @@ public static IHtmlContent Editor(this IHtmlHelper htmlHelper, ModelExpression m
+    /// <param name="newHtmlAttributesObject">new values</param>
+    /// <param name="existingHtmlAttributesObject">existing values</param>
+    /// <returns></returns>
+    internal static IDictionary<string, object> MergeHtmlAttributes(this IHtmlHelper helper, object newHtmlAttributesObject, object existingHtmlAttributesObject)
```
I haven't reviewed this fully but at first glance it seems quite a bit larger than a similar method in the ASP.NET Core itself that does something very similar: https://source.dot.net/#Microsoft.AspNetCore.Mvc.ViewFeatures/DefaultEditorTemplates.cs,192

Could it be simplified?

MSingh-13 replied:
I simplified HtmlHelperExtensions.MergeHtmlAttributes().
See commit 60a82602b1d8a4172cc05a126f7da544662e366e

MSingh-13 committed:
Added a condition to guard against RequiredRoles and RequiredPolicy being set at the same time in AuthzTagHelper.

MSingh-13 committed:
Renamed the properties and attributes related to permission, to resource in AuthzTagHelper.

Changed the type of RequiredPolicyResource property from string to object.
Added a condition to ensure RequiredPolicy is set when RequiredPolicyResource is set.

MSingh-13 committed:
Simplified HtmlHelperExtensions.MergeHtmlAttributes()

MSingh-13 committed:
Refactored MergeHtmlAttributes method.

Further simplified the method.
Correctly overwriting the values of attributes other than class and style.
Renamed the method to MergeHtmlAttributesObjects
Moved it to a new class PublicHtmlHelperExtensions so that it could be made public and used in view templates.

MSingh-13 committed:
Updated DisplayTagHelper and EditorTagHelper.

Switched from an anonymous object to dictionary for htmlAttributes, so that only properties that have values can be added to the dictionary.
Added Id property to set the id on the html element.

Updated samples to demonstrate the merging of class, style and id supplied with editor tag.
Also demonstrates the use for view-data-htmAttributes.

MSingh-13 commented:
Added a new commit to further simplify the MergeHtmlAttributes method. I moved it to a new class in order to make it publicly accessible so that it can be used in templates.

I updated the DisplayTagHelper and EditorTagHelper to include an Id property.
I also updated the samples to demo the merging of htmlAttributes.
